core/environment/fsm_test.go

```diff
@@ -0,0 +1,288 @@
+				Expect(env.Sm.Cannot("START_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("STOP_ACTIVITY")).To(BeTrue())
+				Expect(env.Sm.Cannot("RECOVER")).To(BeTrue())
+				Expect(env.Sm.Cannot("EXIT")).To(BeTrue())
+			})
+		})
+
+	})
+
+	Describe("calling hooks on FSM events", func() {
+		var env *Environment
+		BeforeEach(func() {
+			envId, err := uid.FromString("2oDvieFrVTi")
+			Expect(err).NotTo(HaveOccurred())
+
+			env, err = newEnvironment(map[string]string{}, envId)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(env).NotTo(BeNil())
+		})
+		It("should execute the requested plugin call without errors", func() {
+			env.workflow = workflow.NewAggregatorRole("root", []workflow.Role{
+				workflow.NewCallRole(
+					"call",
+					task.Traits{Trigger: "before_CONFIGURE", Timeout: "5s", Critical: true, Await: "before_CONFIGURE"},
+					"testplugin.Test()",
+					"")})
+			workflow.LinkChildrenToParents(env.workflow)
+			env.Sm.SetState("DEPLOYED")
+			env.workflow.GetUserVars().Del("root.call_called")
+
+			err := env.Sm.Event(context.Background(), "CONFIGURE", NewDummyTransition("CONFIGURE"))
+
+			Expect(err).NotTo(HaveOccurred())
+			v, ok := env.workflow.GetUserVars().Get("root.call_called")
+			Expect(ok).To(BeTrue())
+			Expect(v).To(Equal("true"))
+		})
+
+		It("should return an error if a critical hook fails", func() {
+			env.workflow = workflow.NewAggregatorRole("root", []workflow.Role{
+				workflow.NewCallRole(
+					"call",
+					task.Traits{Trigger: "before_CONFIGURE", Timeout: "5s", Critical: true, Await: "before_CONFIGURE"},
+					"testplugin.Test()",
+					"")})
+			workflow.LinkChildrenToParents(env.workflow)
+			env.Sm.SetState("DEPLOYED")
+			env.workflow.GetUserVars().Del("root.call_called")
+			env.workflow.GetUserVars().Set("testplugin_fail", "true")
+
+			err := env.Sm.Event(context.Background(), "CONFIGURE", NewDummyTransition("CONFIGURE"))
+
+			Expect(err).To(HaveOccurred())
+			v, ok := env.workflow.GetUserVars().Get("root.call_called")
+			Expect(ok).To(BeTrue())
+			Expect(v).To(Equal("true"))
+		})
+
+		It("should not return an error if an non-critical hook fails", func() {
+			env.workflow = workflow.NewAggregatorRole("root", []workflow.Role{
+				workflow.NewCallRole(
+					"call",
+					task.Traits{Trigger: "before_CONFIGURE", Timeout: "5s", Critical: false, Await: "before_CONFIGURE"},
+					"testplugin.Test()",
+					"")})
+			workflow.LinkChildrenToParents(env.workflow)
+			env.Sm.SetState("DEPLOYED")
+			env.workflow.GetUserVars().Del("root.call_called")
+			env.workflow.GetUserVars().Set("testplugin_fail", "true")
+
+			err := env.Sm.Event(context.Background(), "CONFIGURE", NewDummyTransition("CONFIGURE"))
+
+			Expect(err).NotTo(HaveOccurred())
+			v, ok := env.workflow.GetUserVars().Get("root.call_called")
+			Expect(ok).To(BeTrue())
+			Expect(v).To(Equal("true"))
+		})
+
+		It("should return an error if a critical hook times out", func(ctx SpecContext) {
```

Comment:
@teo could please cross-check this case and tell me if I am missing something here? It fails for some reason - the call hangs infinitely and the timeout mechanism does not kick in (I added a `SpecTimeout` to avoid having CI hang).